# Project 3 -- Aman Patel 

# Question 1

In [18]:
video_games <- read.csv("/anvil/projects/tdm/data/grouping/vgchartz-2024.csv")

In [19]:
sort(tapply(video_games$critic_score, video_games$genre, mean, na.rm=TRUE), decreasing=TRUE)

Sandbox     Visual Novel            Music              MMO 
        9.200000         8.500000         7.992593         7.781818 
       Education Action-Adventure     Role-Playing         Strategy 
        7.750000         7.678378         7.470920         7.431733 
        Fighting        Adventure         Platform           Sports 
        7.392424         7.341058         7.307638         7.291027 
         Shooter           Puzzle           Racing       Simulation 
        7.276800         7.201049         7.187524         6.914719 
            Misc           Action            Party       Board Game 
        6.884593         6.831008         5.968750         5.300000

The video game genre with the highest average critic score is Sandbox

In [20]:
sort(tapply(video_games$total_sales, video_games$console, sum, na.rm=TRUE), decreasing=TRUE)

PS2    X360     PS3      PS     PS4     Wii      DS    XOne     PSP      XB 
1027.76  859.79  839.70  546.25  539.92  459.44  458.17  268.96  245.29  232.05 
    GBA      PC      GC     3DS     N64    2600    SNES     PSV     NES      NS 
 224.48  168.99  119.53   99.27   93.79   75.66   65.71   63.02   47.93   36.46 
   WiiU     SAT      GB     GEN      DC     GBC     PSN      NG      WS     SCD 
  35.42   33.67   19.84   18.83   13.31    4.34    3.81    1.44    1.43    0.37 
     VC      WW     XBL     3DO     PCE      GG     Mob     OSX    PCFX    5200 
   0.35    0.25    0.20    0.19    0.16    0.04    0.03    0.03    0.03    0.00 
   7800    ACPC      AJ     AST     Aco     All    Amig     And    ApII     Arc 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
   BBCM     BRW    C128     C64    CD32     CDi      CV     DSi    DSiW     FDS 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
    FMT     GIZ     Int   Linux    Lynx      MS     MSD     MSX   NGage      OR 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
   Ouya     PS5    S32X  Series    TG16      VB    WinP      XS     ZXS     iOS 
   0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00 
   iQue 
   0.00

The PS2 has the highest number of total sales within the console genre

In [21]:
sort(tapply(video_games$title, video_games$genre, length), decreasing=TRUE)

Misc           Action        Adventure     Role-Playing 
            9304             8557             6260             5721 
          Sports          Shooter         Platform         Strategy 
            5586             5410             4001             3685 
          Puzzle           Racing       Simulation         Fighting 
            3521             3425             3158             2367 
Action-Adventure     Visual Novel            Music            Party 
            1877              493              297              151 
             MMO        Education       Board Game          Sandbox 
             115               35               33               20

The function above shows how many games are in each genre from highest to lowest

The tapply() function takes a column, splits the column into groups, and applys an operation to it. This operation could be mean, sum, length,etc. The function is useful with datasets because it allows you to easily split columns and do an action with them rather than having to manually do it with long code segments.

# Question 2

In [6]:
library(dplyr)

In [24]:
video_games %>%
    group_by(genre) %>%
    summarize(avg_critic_score = mean(critic_score, na.rm=TRUE))

genre            avg_critic_score
1  Action           6.831008        
2  Action-Adventure 7.678378        
3  Adventure        7.341058        
4  Board Game       5.300000        
5  Education        7.750000        
6  Fighting         7.392424        
7  MMO              7.781818        
8  Misc             6.884593        
9  Music            7.992593        
10 Party            5.968750        
11 Platform         7.307638        
12 Puzzle           7.201049        
13 Racing           7.187524        
14 Role-Playing     7.470920        
15 Sandbox          9.200000        
16 Shooter          7.276800        
17 Simulation       6.914719        
18 Sports           7.291027        
19 Strategy         7.431733        
20 Visual Novel     8.500000

1 key difference between tapply() and dplyr is that dplyr put the data into a nice table format, while tapply() just lists the data out, allowing for more readability from dplyr. Another example of the difference between the two is that dplyr outputs data frames, while tapply() outputs vectors.

I prefer dplyr simply because it is more easily readable then tapply(). On top of that, for dplyr, you write less, more concise code while printing out the same result

# Question 3

In [25]:
console_sales <- video_games %>%
    group_by(console) %>%
    summarize(sum_sales = sum(total_sales, na.rm=TRUE)) %>%
    arrange(desc(sum_sales))
    head(console_sales, n=20)

console sum_sales
1  PS2     1027.76  
2  X360     859.79  
3  PS3      839.70  
4  PS       546.25  
5  PS4      539.92  
6  Wii      459.44  
7  DS       458.17  
8  XOne     268.96  
9  PSP      245.29  
10 XB       232.05  
11 GBA      224.48  
12 PC       168.99  
13 GC       119.53  
14 3DS       99.27  
15 N64       93.79  
16 2600      75.66  
17 SNES      65.71  
18 PSV       63.02  
19 NES       47.93  
20 NS        36.46

PS2 has the most sales with $1027.76

In [26]:
genre_title_counts <- video_games %>%
    group_by(genre) %>%
    summarize(total_titles = sum(n(), na.rm=TRUE)) %>%
    arrange(desc(total_titles)) 

head(genre_title_counts)

genre        total_titles
1 Misc         9304        
2 Action       8557        
3 Adventure    6260        
4 Role-Playing 5721        
5 Sports       5586        
6 Shooter      5410

Shooter games had the least title counts with only 5410

In [27]:
high_sold_games <- video_games %>%
    group_by(title) %>%
    summarize(high_sell = sum(total_sales, na.rm=TRUE)) %>%
    arrange(desc(high_sell)) 

head(high_sold_games, n=10)

title                          high_sell
1  Grand Theft Auto V             64.29    
2  Call of Duty: Black Ops        30.99    
3  Call of Duty: Modern Warfare 3 30.71    
4  Call of Duty: Black Ops II     29.59    
5  Call of Duty: Ghosts           28.80    
6  Call of Duty: Black Ops 3      26.72    
7  Call of Duty: Modern Warfare 2 25.02    
8  Minecraft                      24.01    
9  Grand Theft Auto IV            22.53    
10 Call of Duty: Advanced Warfare 21.78

The table above shows the top 10 most sold games (in millions), with Grand Theft Auto having sold 64.29 million copies

# Question 4

In [2]:
app_store <- read.csv("/anvil/projects/tdm/data/grouping/googleplaystore.csv")

In [7]:
app_rating <- app_store %>%
    group_by(Rating) %>%
    summarize(total_reviews = sum(n(), na.rm=TRUE)) %>%
    arrange(desc(total_reviews))

head(app_rating)

Rating total_reviews
1 NaN    1474         
2 4.4    1109         
3 4.3    1076         
4 4.5    1038         
5 4.2     952         
6 4.6     823

In [13]:
app_store$Clean_Reviews <- gsub(',','',app_store$Review)
app_store$Numeric_Reviews <- as.numeric(app_store$Clean_Reviews)
app_store[is.na(app_store$Numeric_Reviews) & !is.na(app_store$Reviews), "Reviews"]

Warning message:
"NAs introduced by coercion"


[1] "3.0M"

In [10]:
app_store_clean <- app_store %>%
    mutate(
        Clean_Reviews = gsub(",", "", Reviews),
        Numeric_Reviews = as.numeric(Clean_Reviews)
    ) %>%
    filter(!is.na(Numeric_Reviews))

Warning message:
"There was 1 warning in `mutate()`.
i In argument: `Numeric_Reviews = as.numeric(Clean_Reviews)`.
Caused by warning:
! NAs introduced by coercion"


mutate() -> adds new columns or it can change existing columns in the dataset

filter() -> keeps only the rows that meet certain conditions; this was used to remove the 3.0M value

gsub() -> this function is used to replace parts of the data. In this dataset, we used gsub() to replace the "," with ""(nothing)

In [14]:
app_store_cleaned <- app_store_clean %>%
    mutate(rating_per_review = Rating / Numeric_Reviews)

head(app_store_cleaned$rating_per_review)

[1] 2.578616e-02 4.033092e-03 5.370815e-05 2.086773e-05 4.446743e-03
[6] 2.634731e-02

# Question 5

lapply() - outputs a list after applying a function to each element of the list

sapply() - applies the function to a list and then simplifies it

In [37]:
lapply(app_store_cleaned, class)

$App
[1] "character"

$Category
[1] "character"

$Rating
[1] "numeric"

$Reviews
[1] "character"

$Size
[1] "character"

$Installs
[1] "character"

$Type
[1] "character"

$Price
[1] "character"

$Content.Rating
[1] "character"

$Genres
[1] "character"

$Last.Updated
[1] "character"

$Current.Ver
[1] "character"

$Android.Ver
[1] "character"

$Clean_Reviews
[1] "character"

$Numeric_Reviews
[1] "numeric"

$rating_per_review
[1] "numeric"

In [18]:
sapply(app_store_cleaned, class)

App          Category            Rating           Reviews 
      "character"       "character"         "numeric"       "character" 
             Size          Installs              Type             Price 
      "character"       "character"       "character"       "character" 
   Content.Rating            Genres      Last.Updated       Current.Ver 
      "character"       "character"       "character"       "character" 
      Android.Ver     Clean_Reviews   Numeric_Reviews rating_per_review 
      "character"       "character"         "numeric"         "numeric"

In [16]:
head(lapply(app_store_cleaned$Content.Rating, nchar), n=30)

[[1]]
[1] 8

[[2]]
[1] 8

[[3]]
[1] 8

[[4]]
[1] 4

[[5]]
[1] 8

[[6]]
[1] 8

[[7]]
[1] 8

[[8]]
[1] 8

[[9]]
[1] 8

[[10]]
[1] 8

[[11]]
[1] 8

[[12]]
[1] 8

[[13]]
[1] 4

[[14]]
[1] 8

[[15]]
[1] 8

[[16]]
[1] 8

[[17]]
[1] 8

[[18]]
[1] 8

[[19]]
[1] 8

[[20]]
[1] 8

[[21]]
[1] 8

[[22]]
[1] 8

[[23]]
[1] 12

[[24]]
[1] 8

[[25]]
[1] 8

[[26]]
[1] 8

[[27]]
[1] 8

[[28]]
[1] 8

[[29]]
[1] 8

[[30]]
[1] 8

In [40]:
head(sapply(app_store_cleaned$Content.Rating, nchar), n=30)

Everyone     Everyone     Everyone         Teen     Everyone     Everyone 
           8            8            8            4            8            8 
    Everyone     Everyone     Everyone     Everyone     Everyone     Everyone 
           8            8            8            8            8            8 
        Teen     Everyone     Everyone     Everyone     Everyone     Everyone 
           4            8            8            8            8            8 
    Everyone     Everyone     Everyone     Everyone Everyone 10+     Everyone 
           8            8            8            8           12            8 
    Everyone     Everyone     Everyone     Everyone     Everyone     Everyone 
           8            8            8            8            8            8

I used this website to research a little bit about the basic apply() function: https://www.dataquest.io/blog/apply-functions-in-r-sapply-lapply-tapply

The basic apply() function is used to apply a function to a dataset. There are 3 main arguments taken: the dataset, the margin, and the function that has to be applied. '1' in the margin argument is rows and '2' is columns. Unlike lapply() or sapply(), which applies functions to lists, apply() applies a function to tables with numeric values. 